cs/src/core/RuntimeDeserializer.cs

```diff
@@ -0,0 +1,671 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Bond
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using Bond.IO.Safe;
+    using Bond.Protocols;
+
+    /// <summary>
+    /// Deserializes <see cref="RuntimeObject"/>s from a protocol reader using only a runtime schema.
+    /// </summary>
+    public static class RuntimeDeserialize
+    {
+        /// <summary>
+        /// Deserializes a <see cref="RuntimeObject"/> from a tagged protocol reader using only a runtime schema.
+        /// </summary>
+        /// <param name="reader">The tagged protocol reader to read the object data from.</param>
+        /// <param name="schema">The schema to use when reading the object data.</param>
+        /// <returns>A <see cref="RuntimeObject"/> created from <paramref name="reader"/> and <paramref name="schema"/>.</returns>
+        public static RuntimeObject From(IClonableTaggedProtocolReader reader, RuntimeSchema schema)
+        {
+            return new RuntimeDeserializer(schema).Deserialize(reader);
+        }
+
+        /// <summary>
+        /// Deserializes a <see cref="RuntimeObject"/> from an untagged protocol reader using only a runtime schema.
+        /// </summary>
+        /// <param name="reader">The untagged protocol reader to read the object data from.</param>
+        /// <param name="schema">The schema to use when reading the object data.</param>
+        /// <returns>A <see cref="RuntimeObject"/> created from <paramref name="reader"/> and <paramref name="schema"/>.</returns>
+        public static RuntimeObject From(IClonableUntaggedProtocolReader reader, RuntimeSchema schema)
+        {
+            return new RuntimeDeserializer(schema).Deserialize(reader);
+        }
+    }
+
+    /// <summary>
+    /// Deserializes <see cref="RuntimeObject"/>s from a protocol reader using only a <see cref="RuntimeSchema"/>.
+    /// </summary>
+    public class RuntimeDeserializer
```

Comment:
I agree that this does not follow how existing Bond deserialization works, but that was out of necessity. In existing code, transforms are created/compiled once on a per-.NET-type basis, but we will often not have corresponding .NET types for the schema we are deserializing. A transform could be created for the provided schema every time a RuntimeDeserializer is created, but that seemed unreasonable.

I do not see an obvious solution for making this work for for all current and future Bond protocols. Explicit support for JSON and XML could be added, but that would require either moving RuntimeSerializer and RuntimeDeserializer to a new project (so it can reference the JSON project) or creating something like RuntimeJsonSerializer and RuntimeJsonDeserializer in the JSON project.